### ISSUE TITLE
NullPointerException for Azure AD JWK definition on missing algorithm ID

### ISSUE THREAD

**Issue by djechelon** [spring-projects/spring-security-oauth#1183]
I am building an Azure-AD bases SSO. I am working on my own as I can't find out-of-the-box examples.

Currently I have build XML definition for an Authentication endpoint using Azure AD and JWT tokens. Please do not mind the abuse of "office365" name

```
    <oauth2:client id="oauth2ClientContextFilter" />


    <!-- Risorsa protetta OAuth per chiamata /me -->
    <oauth2:resource id="office365AuthenticationSource" client-id="${security.oauth.appId}" client-secret="${security.oauth.appSecret}" type="authorization_code"
        authentication-scheme="form" client-authentication-scheme="form" use-current-uri="true" user-authorization-uri="${security.oauth.authorizationUri}" access-token-uri="${security.oauth.accessTokenUri}"
        scope="${security.oauth.scopes}" pre-established-redirect-uri="${security.oauth.redirectUri}" />


    <oauth2:rest-template id="office365OAuthRestTemplate" resource="office365AuthenticationSource" />

    <b:bean id="office365UserDetailsService" class="it.web.security.spring.Office365UserDetailsService" />

    <b:bean id="oauth2ClientAuthenticationEntryPointFilter" class="org.springframework.security.oauth2.client.filter.OAuth2ClientAuthenticationProcessingFilter">
        <b:constructor-arg value="/auth/office365" />
        <b:property name="restTemplate" ref="office365OAuthRestTemplate" />
        <b:property name="tokenServices">
            <b:bean class="org.springframework.security.oauth2.provider.token.DefaultTokenServices">
                <b:property name="tokenStore">
                    <b:bean class="org.springframework.security.oauth2.provider.token.store.jwk.JwkTokenStore">
                        <b:constructor-arg value="${security.oauth.jwkStore}" />
                    </b:bean>
                </b:property>
            </b:bean>
        </b:property>
    </b:bean>

```

Noteworthy: the client authentication processing filter invokes Azure AD (misnamed Office 365) to get a token scoped for `openid` which is then passed to the JWK token store. The store loads keys from `https://login.microsoftonline.com/common/discovery/v2.0/keys`. It happens that MS discovery document is missing the `alg` parameter, resulting in a NPE.

```
java.lang.NullPointerException
	at org.springframework.security.oauth2.provider.token.store.jwk.JwkVerifyingJwtAccessTokenConverter.decode(JwkVerifyingJwtAccessTokenConverter.java:114)
	at org.springframework.security.oauth2.provider.token.store.JwtTokenStore.convertAccessToken(JwtTokenStore.java:88)
	at org.springframework.security.oauth2.provider.token.store.JwtTokenStore.readAccessToken(JwtTokenStore.java:80)
	at org.springframework.security.oauth2.provider.token.store.jwk.JwkTokenStore.readAccessToken(JwkTokenStore.java:212)
	at org.springframework.security.oauth2.provider.token.DefaultTokenServices.loadAuthentication(DefaultTokenServices.java:229)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.__invoke(DelegatingMethodAccessorImpl.java:43)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:333)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:207)
	at com.sun.proxy.$Proxy167.loadAuthentication(Unknown Source)
	at org.springframework.security.oauth2.client.filter.OAuth2ClientAuthenticationProcessingFilter.attemptAuthentication(OAuth2ClientAuthenticationProcessingFilter.java:112)
```

Looking at Spring implementation, scanning the JWK discovery file does not manage the absence of the algorithm key in the response, either by setting a default or throwing an exception. Nor the verifier in the offending line handles the case where algorithm of the key is null.

Offending [code](https://github.com/spring-projects/spring-security-oauth/blob/2.2.0.RELEASE/spring-security-oauth2/src/main/java/org/springframework/security/oauth2/provider/token/store/jwk/JwkVerifyingJwtAccessTokenConverter.java#L114)

Related  #849

**Comment by djechelon** [comment]
Update, I did a little reseach

According to RFC7517 "alg" is _OPTIONAL_, but from the RFCs themselves it looks like "alg" may be required according to the value of "kty", so this suggests Microsoft could be doing the wrong thing.

Really, _can_ MS do the wrong thing? For the record [Google](https://www.googleapis.com/oauth2/v3/certs) uses `RS256` explicitly

**Comment by vpavic** [comment]
@djechelon See #1129 and #1130 - this has already been fixed and will be a part of upcoming `2.2.1.RELEASE`.

**Comment by djechelon** [comment]
Sorry for the dupes. Closing
